File: Roadmap/10 - EXCEPCIONES/c++/N0HagoNada.cpp
#include <iostream>
#include <vector>
#include <stdexcept>

double division(double a, double b)
{
    if (b == 0)
    {
        throw std::runtime_error("Error: División por cero");
    }
    return a / b;
}

int obtenerValorIndice(const std::vector<int> &numeros, int indice)
{
    if (indice < 0 || indice >= static_cast<int>(numeros.size()))
    {
        throw std::out_of_range("Error: Índice fuera de rango");
    }
    return numeros[indice];
}

int main()
{
    try
    {
        double resultado = division(10, 0);
        std::cout << "Resultado: " << resultado << std::endl;
    }
    catch (const std::runtime_error &e)
    {
        std::cerr << e.what() << std::endl;
    }

    try
    {
        std::vector<int> numeros = {1, 2, 3};
        int valor = obtenerValorIndice(numeros, 5);
        std::cout << "Valor: " << valor << std::endl;
    }
    catch (const std::out_of_range &e)
    {
        std::cerr << e.what() << std::endl;
    }

    std::cout << "Fin del programa" << std::endl;
    return 0;
}

File: Roadmap/10 - EXCEPCIONES/go/N0HagoNada.go
package main

import (
	"errors"
	"fmt"
)

type CustomError struct {
	Message string
}

func (e *CustomError) Error() string {
	return e.Message
}

func main() {
	// Ejemplo 1: División por cero
	result, err := divideNumbers(10, 0)
	if err != nil {
		fmt.Println("Error:", err)
	} else {
		fmt.Println("Resultado:", result)
	}

	// Ejemplo 2: Acceso a índice fuera de rango
	numbers := []int{1, 2, 3}
	value, err := getValueAtIndex(numbers, 5)
	if err != nil {
		fmt.Println("Error:", err)
	} else {
		fmt.Println("Valor:", value)
	}
	fmt.Println("****************************** Reto ***************************************")
	// Llamada a la función processParams con diferentes casos
	err = processParams(10, 2)
	handleError(err)

	err = processParams(0, 5)
	handleError(err)

	err = processParams(10, 0)
	handleError(err)

	err = processParams(-5, 3)
	handleError(err)

	err = processParams(8, 4)
	handleError(err)

	fmt.Println("Ejecución finalizada")
}

func divideNumbers(a, b int) (int, error) {
	if b == 0 {
		return 0, fmt.Errorf("No se puede dividir por cero")
	}
	return a / b, nil
}

func getValueAtIndex(numbers []int, index int) (int, error) {

	if index < 0 || index >= len(numbers) {
		return 0, fmt.Errorf("Índice fuera de rango: %d", index)
	}

	return numbers[index], nil
}

// Función que procesa parámetros y puede lanzar diferentes tipos de excepciones
func processParams(a, b int) error {
	if a == 0 {
		return errors.New("No se puede procesar con el valor 'a' igual a cero")
	}

	if b == 0 {
		return fmt.Errorf("No se puede procesar con el valor 'b' igual a cero")
	}

	if a < 0 || b < 0 {
		return &CustomError{Message: "No se pueden procesar valores negativos"}
	}

	// Procesamiento correcto
	result := a / b
	fmt.Printf("Resultado: %d\n", result)
	return nil
}

func handleError(err error) {
	if err != nil {
		fmt.Println("Tipo de error:", err)
	} else {
		fmt.Println("No se ha producido ningún error")
	}
}
